### ISSUE TITLE
Support forward pre-hooks for modules in TorchScript

### ISSUE THREAD

**Issue by yf225** [pytorch/pytorch#42600]
As part of the PyPer model TorchScript conversion work, we introduced a concept called "placeholder module" which has the same forward signature as the actual module, but only some of them are replaced with the actual module in the lazy initialization step before the parent module's first forward call. Here is a minimal example:

```python
class TestModule(torch.nn.Module):
    def __init__(self):
        super(TestModule, self).__init__()
        self.mod_a = create_placeholder_module(ModuleA)
        self.mod_b = create_placeholder_module(ModuleB)

    def init_params(self, inputs):
        if inputs.shape[0] == 1:
            self.mod_a = ModuleA(inputs.shape[1])  # Use input to create actual module
        elif inputs.shape[0] == 2:
            self.mod_b = ModuleB(inputs.shape[1])  # Use input to create actual module

    def forward(self, inputs):
        if inputs.shape[0] == 1:
            return self.mod_a(inputs)
        elif inputs.shape[0] == 2:
            return self.mod_b(inputs)

m = TestModule()
m.init_params(torch.randn(1, 3))  # only `self.mod_a` is initialized
m_scripted = torch.jit.script(m)

m_scripted(torch.randn(1, 3))  # behave as expected
m_scripted(torch.randn(2, 3))  # TODO: we want to throw error in this case
```

This design works for us, but we do want to error out when user calls into a placeholder module (i.e. the `inputs.shape[0] == 2` branch). We believe one way to support this is to add a forward pre-hook to all placeholder modules, and we are curious about the progress on https://github.com/pytorch/pytorch/issues/29037 and https://github.com/pytorch/pytorch/pull/36183. **This will make PyPer scripted models much easier to use and gives better error messages when users pass in an unsupported input**.

Thanks!

cc. @wanchaol @suo

cc @suo @gmagogsfm

**Comment by gmagogsfm** [comment]
I am probably missing some context, but curios why you don't choose to make placeholder modules with a dummy forward that throws an exception?

**Comment by yf225** [comment]
> I am probably missing some context, but curios why you don't choose to make placeholder modules with a dummy forward that throws an exception?

It's because the placeholder module still needs to have the same forward method signature as the actual module (otherwise it can't be a placeholder for TS compilation purpose). For example if `ModuleA` has `def forward(self, input1: Tensor, input2: Tensor)`, then `ModuleAPlaceholderModule` also need to have `def forward(self, input1: Tensor, input2: Tensor)`. The only scalable way to generate this placeholder module class for every module class is probably something like https://www.internalfb.com/intern/diffusion/FBS/browse/master/fbcode/dper3/dper3/core/pytorch_export_utils.py?commit=1f1fb1fe7fd426752bd055de32e32021a615eb25&lines=129%2C130%2C131%2C132 (i.e. subclass from the actual module class), but then we don't have the opportunity to change the `forward` implementation`.

**Comment by Lilyjjo** [comment]
Closed with PR #49544
